racket/src/thread/future.rkt

```diff
@@ -0,0 +1,441 @@
+#lang racket/base
+
+(require "check.rkt"
+         "internal-error.rkt"
+         "engine.rkt"
+         "atomic.rkt"
+         "parameter.rkt"
+         "../common/queue.rkt"
+         "thread.rkt")
+
+(provide futures-enabled?
+         processor-count
+         current-future
+         future
+         future?
+         would-be-future
+         touch
+         future-block
+         current-future-prompt
+         future:condition-broadcast
+         future:condition-signal
+         future:condition-wait
+         future:make-condition
+         lock-release
+         lock-acquire
+         make-lock
+         with-lock
+         )
+;; ---------------------------- locks ------------------------------------
+
+(define-syntax-rule (with-lock lock expr ...)
+  (begin
+    (lock-acquire lock)
+    (begin0
+        (let () expr ...)
+      (lock-release lock))))
+
+(struct future-lock* (box owner))
+
+(define (make-lock)
+  (future-lock* (box 0) (box #f)))
+
+;;  cannot use atomics in locking code because it is used by code that
+;; isn't running in engines.
+;; would have to check whether running in an engine to be atomic.
+;; don't think its necessary though
+(define (lock-acquire lock [block? #t])
+  (define box (future-lock*-box lock))
+  (let loop ()
+    (cond
+      [(and (= 0 (unbox box)) (box-cas! box 0 1)) ;; got lock
+       (let ([caller (get-caller)])
+         (unless (box-cas! (future-lock*-owner lock) #f caller)
+           (internal-error "Lock already has owner."))
+         #t)]
+      [block?
+       (loop)]
+      [else
+       #f])))
+
+(define (lock-release lock)
+  (define caller (get-caller))
+  (when (eq? caller (unbox (future-lock*-owner lock)))
+    (unless (box-cas! (future-lock*-owner lock) caller #f)
+      (internal-error "Failed to reset owner\n"))
+    (unless (box-cas! (future-lock*-box lock) 1 0)
+      (internal-error "Lock release failed\n"))))
+
+(define (own-lock? lock caller)
+  (and (eq? caller (unbox (future-lock*-owner lock)))
+       (begin0 #t
+         (unless (= 1 (unbox (future-lock*-box lock)))
+           (internal-error "Caller 'owns' lock but lock is free.")))))
+
+;; not sure of order here...
+(define (get-caller)
+  (cond
+    [(current-future)
+     (current-future)]
+    [(not (= 0 (get-pthread-id)))
+     (get-pthread-id)]
+    [else
+     (current-thread)]))
+
+
+;; ---------------------------- futures ----------------------------------
+
+
+(define ID (box 1))
+
+(define get-next-id
+  (lambda ()
+    (let ([id (unbox ID)])
+      (if (box-cas! ID id (+ 1 id))
+          id
+          (get-next-id)))))
+
+(define (processor-count)
+  1)
+
+(define futures-enabled? threaded?)
+
+(struct future* (id cond lock prompt
```

Comment:
I think `#:auto` has turned out to be a bad idea. Instead of using `#:auto`, I recommend having a separate `create-future` function that is always used instead of the main constructor (and that fills in all the would-be auto fields with suitable initial values).